File: malai-core/org.malai.ts/src-core/fsm/FSM.ts
/*
 * This file is part of Malai.
 * Copyright (c) 2009-2018 Arnaud BLOUIN
 * Malai is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * Malai is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */

import {State} from "./State";
import {InitState} from "./InitState";
import {ObsValue} from "../utils/ObsValue";
import {OutputState} from "./OutputState";
import {FSMHandler} from "./FSMHandler";
import {TimeoutTransition} from "./TimeoutTransition";
import {StdState} from "./StdState";
import {InputState} from "./InputState";
import {Logger} from "typescript-logging";
import {factory} from "../logging/ConfigLog";
import {MArray} from "../../src/util/ArrayUtil";
import {OutputStateImpl} from "./OutputStateImpl";

export class FSM<E> {
    protected logger: Logger | undefined;

    protected _inner: boolean;

    /**
     * By default an FSM triggers its 'start' event when it leaves its initial state.
     * In some cases, this is not the case. For example, a double-click interaction is an FSM that must trigger
     * its start event when the FSM reaches... its terminal state. Similarly, a DnD must trigger its start event
     * on the first move, not on the first press.
     * The goal of this attribute is to identify the state of the FSM that must trigger the start event.
     * By default, this attribute is set with the initial state of the FSM.
     */
    protected _startingState: State<E>;

    /**
     * Goes with 'startingState'. It permits to know whether the FSM has started, ie whether the 'starting state'
     * has been reached.
     */
    protected started: boolean;

    public readonly initState: InitState<E>;

    protected readonly _currentState: ObsValue<OutputState<E>>;

    /**
     * The states that compose the finite state machine.
     */
    protected readonly states: MArray<State<E>>;

    /**
     * The handler that want to be notified when the state machine of the interaction changed.
     */
    protected readonly handlers: MArray<FSMHandler>;

    /**
     * The events still in process. For example when the user press key ctrl and scroll one time using
     * the wheel of the mouse, the interaction scrolling is
     * finished but the event keyPressed 'ctrl' is still in process. At the end of the interaction, these events
     * are re-introduced into the
     * state machine of the interaction for processing.
     */
    protected readonly eventsToProcess: MArray<E>;

    /**
     * The current timeout in progress.
     */
    protected currentTimeout: TimeoutTransition<E> | undefined;

    protected currentSubFSM: FSM<E> | undefined;

    public constructor() {
        this._inner = false;
        this.started = false;
        this.started = false;
        this.initState = new InitState<E>(this, "init");
        this.states = new MArray<State<E>>(this.initState);
        this._startingState = this.initState;
        this._currentState = new ObsValue<OutputState<E>>(this.initState);
        this.handlers = new MArray();
        this.eventsToProcess = new MArray();
    }

    public setCurrentSubFSM(subFSM?: FSM<E>): void {
        this.currentSubFSM = subFSM;
    }

    public get currentState(): OutputState<E> {
        return this._currentState.get();
    }

    public set inner(inner: boolean) {
        this._inner = inner;
    }

    public get inner(): boolean {
        return this._inner;
    }

    public process(event: E): boolean {
        if (event === undefined) {
            return false;
        }
        if (this.currentSubFSM !== undefined) {
            return this.currentSubFSM.process(event);
        }
        return this._currentState.get().process(event);
    }

    public enterStdState(state: StdState<E>): void {
        this.currentState = state;
        this.checkTimeoutTransition();
        if (this.started) {
            this.onUpdating();
        }
    }

    public isStarted(): boolean {
        return this.started;
    }

    public set currentState(state: OutputState<E>) {
        this._currentState.set(state);
    }

    /**
     * At the end of the FSM execution, the events still (eg keyPress) in process must be recycled to be reused in the FSM.
     */
    protected processRemainingEvents(): void {
        if (this.eventsToProcess !== undefined) {
            const list: MArray<E> = new MArray(...this.eventsToProcess);
            while (list.length > 0) {
                const event = list.removeAt(0);

                if (event) {
                    this.eventsToProcess.removeAt(0);
                    if (this.logger !== undefined) {
                        this.logger.info("Recycling event: " + String(event));
                    }
                    this.process(event);
                }
            }
        }
    }

    public addRemaningEventsToProcess(event: E): void {
        if (event !== undefined) {
            this.eventsToProcess.push(event);
        }
    }

    /**
     * Terminates the state machine.
     */
    public onTerminating(): void {
        if (this.logger !== undefined) {
            this.logger.info("FSM ended");
        }
        if (this.started) {
            this.notifyHandlerOnStop();
        }
        this.reinit();
        this.processRemainingEvents();
    }

    /**
     * Cancels the state machine.
     */
    public onCancelling(): void {
        if (this.logger !== undefined) {
            this.logger.info("FSM cancelled");
        }
        if (this.started) {
            this.notifyHandlerOnCancel();
        }
        this.fullReinit();
    }

    /**
     * Starts the state machine.
     */
    public onStarting(): void {
        if (this.logger !== undefined) {
            this.logger.info("FSM started");
        }
        this.started = true;
        this.notifyHandlerOnStart();
    }

    /**
     * Updates the state machine.
     */
    public onUpdating(): void {
        if (this.started) {
            if (this.logger !== undefined) {
                this.logger.info("FSM updated");
            }
            this.notifyHandlerOnUpdate();
        }
    }

    /**
     * Adds a state to the state machine.
     * @param {*} state The state to add. Must not be null.
     */
    public addState(state: InputState<E>): void {
        if (state !== undefined) {
            this.states.push(state);
        }
    }

    public log(log: boolean): void {
        if (log) {
            if (this.logger === undefined) {
                this.logger = factory.getLogger("FSM");
            }
        } else {
            this.logger = undefined;
        }
    }

    public reinit(): void {
        if (this.logger !== undefined) {
            this.logger.info("FSM reinitialised");
        }
        if (this.currentTimeout !== undefined) {
            this.currentTimeout.stopTimeout();
        }
        this.started = false;
        this._currentState.set(this.initState);
        this.currentTimeout = undefined;
        if (this.currentSubFSM !== undefined) {
            this.currentSubFSM.reinit();
        }
    }

    public fullReinit(): void {
        if (this.eventsToProcess !== undefined) {
            this.eventsToProcess.clear();
        }
        this.reinit();
        if (this.currentSubFSM !== undefined) {
            this.currentSubFSM.fullReinit();
        }
    }

    public onTimeout() {
        if (this.currentTimeout !== undefined) {
            if (this.logger !== undefined) {
                this.logger.info("Timeout");
            }
            const state = this.currentTimeout.execute().get();
            if (state instanceof OutputStateImpl) {
                this._currentState.set(state);
                this.checkTimeoutTransition();
            }
        }
    }

    /**
     * Stops the current timeout transition.
     */
    public stopCurrentTimeout(): void {
        if (this.currentTimeout !== undefined) {
            if (this.logger !== undefined) {
                this.logger.info("Timeout stopped");
            }
            this.currentTimeout.stopTimeout();
            this.currentTimeout = undefined;
        }
    }

    /**
     * Checks whether the current state has a timeout transition.
     * If it is the case, the timeout transition is launched.
     */
    protected checkTimeoutTransition(): void {
        const tr = this._currentState.get().getTransitions().find(t => t instanceof TimeoutTransition) as TimeoutTransition<E> | undefined;

        if (tr) {
            if (this.logger !== undefined) {
                this.logger.info("Timeout starting");
            }
            this.currentTimeout = tr;
            this.currentTimeout.startTimeout();
        }
    }

    public addHandler(handler: FSMHandler): void {
        if (handler !== undefined) {
            this.handlers.push(handler);
        }
    }

    public removeHandler(handler: FSMHandler): void {
        if (handler !== undefined) {
            this.handlers.remove(handler);
        }
    }

    /**
     * Notifies handler that the interaction starts.
     */
    protected notifyHandlerOnStart(): void {
        try {
            this.handlers.forEach(handler => handler.fsmStarts());
        } catch (ex) {
            this.onCancelling();
            throw ex;
        }
    }

    /**
     * Notifies handler that the interaction updates.
     */
    protected notifyHandlerOnUpdate(): void {
        try {
            this.handlers.forEach(handler => handler.fsmUpdates());
        } catch (ex) {
            this.onCancelling();
            throw ex;
        }
    }

    /**
     * Notifies handler that the interaction stops.
     */
    public notifyHandlerOnStop() {
        try {
            this.handlers.forEach(handler => handler.fsmStops());
        } catch (ex) {
            this.onCancelling();
            throw ex;
        }
    }

    /**
     * Notifies handler that the interaction is cancelled.
     */
    protected notifyHandlerOnCancel(): void {
        this.handlers.forEach(handler => handler.fsmCancels());
    }

    public getStates(): Array<State<E>> {
        return [...this.states];
    }

    public currentStateProp(): ObsValue<OutputState<E>> {
        return this._currentState;
    }

    public get startingState(): State<E> {
        return this._startingState;
    }

    public set startingState(state: State<E>) {
        if (state !== undefined) {
            this._startingState = state;
        }
    }

    public getEventsToProcess(): MArray<E> {
        return new MArray(...this.eventsToProcess);
    }
}

File: src/main/java/io/github/interacto/fsm/FSM.java
/*
 * Interacto
 * Copyright (C) 2020 Arnaud Blouin
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package io.github.interacto.fsm;

import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * A finite state machine that defines the behavior of a user interaction.
 * @param <E> The type of events the FSM processes.
 */
public class FSM<E> {
	protected Logger logger;
	protected boolean inner;
	/**
	 * By default an FSM triggers its 'start' event when it leaves its initial state.
	 * In some cases, this is not the case. For example, a double-click interaction is an FSM that must trigger
	 * its start event when the FSM reaches... its terminal state. Similarly, a DnD must trigger its start event
	 * on the first move, not on the first press.
	 * The goal of this attribute is to identify the state of the FSM that must trigger the start event.
	 * By default, this attribute is set with the initial state of the FSM.
	 */
	protected State<E> startingState;
	/** Goes with 'startingState'. It permits to know whether the FSM has started, ie whether the 'starting state' has been reached. */
	protected boolean started;
	protected final InitState<E> initState;
	protected OutputState<E> currentState;
	protected final PublishSubject<Map.Entry<OutputState<E>, OutputState<E>>> currentStatePublisher;
	/** The states that compose the finite state machine. */
	protected final Set<State<E>> states;
	/** The handler that want to be notified when the state machine of the interaction changed. */
	protected final Set<FSMHandler> handlers;
	/**
	 * The events still in process. For example when the user press key ctrl and scroll one time using the wheel of the mouse, the interaction scrolling is
	 * finished but the event keyPressed 'ctrl' is still in process. At the end of the interaction, these events are re-introduced into the
	 * state machine of the interaction for processing.
	 */
	protected final List<E> eventsToProcess;
	/** The current timeout in progress. */
	protected TimeoutTransition<E> currentTimeout;
	protected FSM<E> currentSubFSM;


	/**
	 * Creates the FSM.
	 */
	public FSM() {
		super();
		eventsToProcess = new ArrayList<>();
		started = false;
		initState = new InitState<>(this, "init");
		states = new HashSet<>();
		states.add(initState);
		startingState = initState;
		currentState = initState;
		currentStatePublisher = PublishSubject.create();
		inner = false;
		handlers = new HashSet<>(2);
	}

	/**
	 * @return The current state of FSM during its execution.
	 */
	public OutputState<E> getCurrentState() {
		return currentState;
	}

	/**
	 * @return An observable value for observing the current state of FSM during its execution.
	 */
	public Observable<Map.Entry<OutputState<E>, OutputState<E>>> currentState() {
		return currentStatePublisher;
	}

	/**
	 * States whether the FSM is an inner FSM (ie, whether it is included into another FSM as
	 * a sub-FSM transition).
	 * @param inner True: this FSM will be considered as an inner FSM.
	 */
	public void setInner(final boolean inner) {
		this.inner = inner;
	}

	/**
	 * @return True: this FSM is an inner FSM.
	 */
	public boolean isInner() {
		return inner;
	}

	/**
	 * Processes the provided event to run the FSM.
	 * @param event The event to process.
	 * @return True: the FSM correctly processed the event.
	 */
	public boolean process(final E event) {
		if(event == null) {
			return false;
		}
		if(currentSubFSM != null) {
			return currentSubFSM.process(event);
		}
		return currentState.process(event);
	}

	protected void enterStdState(final StdState<E> state) throws CancelFSMException {
		setCurrentState(state);
		checkTimeoutTransition();
		if(started) {
			onUpdating();
		}
	}

	/**
	 * @return True: The FSM started.
	 */
	public boolean isStarted() {
		return started;
	}

	protected void setCurrentState(final OutputState<E> state) {
		final var old = currentState;
		currentState = state;
		currentStatePublisher.onNext(Map.entry(old, currentState));
	}

	/**
	 * At the end of the FSM execution, the events still (eg keyPress) in process must be recycled to be reused in the FSM.
	 */
	protected void processRemainingEvents() {
		synchronized(eventsToProcess) {
			// All the events must be processed but the list stillProcessingEvents can be modified
			// during the process. So, a clone of the list must be created.
			final List<E> list = new ArrayList<>(eventsToProcess);

			// All the events must be processed.
			while(!list.isEmpty()) {
				final E event = list.remove(0);
				// Do not forget to remove the event from its original list.
				eventsToProcess.remove(0);

				if(logger != null) {
					logger.log(Level.INFO, "Recycling event: " + event);
				}

				process(event);
			}
		}
	}

	protected void addRemaningEventsToProcess(final E event) {
		if(event != null) {
			synchronized(eventsToProcess) {
				eventsToProcess.add(event);
			}
		}
	}

	/**
	 * Terminates the state machine.
	 * @throws CancelFSMException If the interaction is cancelled by a handler during the stopping step.
	 */
	protected void onTerminating() throws CancelFSMException {
		if(logger != null) {
			logger.log(Level.INFO, "FSM ended");
		}

		if(started) {
			notifyHandlerOnStop();
		}
		reinit();
		processRemainingEvents();
	}

	/**
	 * Cancels the state machine.
	 * */
	protected void onCancelling() {
		if(logger != null) {
			logger.log(Level.INFO, "FSM cancelled");
		}

		if(started) {
			notifyHandlerOnCancel();
		}
		// When an interaction is aborted, the events in progress must not be reused.
		fullReinit();
	}

	/**
	 * Starts the state machine.
	 * @throws CancelFSMException If the interaction is cancelled by a handler during the starting step.
	 */
	public void onStarting() throws CancelFSMException {
		if(logger != null) {
			logger.log(Level.INFO, "FSM started");
		}

		started = true;
		notifyHandlerOnStart();
	}

	/**
	 * Updates the state machine.
	 * @throws CancelFSMException If the interaction is cancelled by a handler during the updating step.
	 */
	public void onUpdating() throws CancelFSMException {
		if(started) {
			if(logger != null) {
				logger.log(Level.INFO, "FSM updated");
			}

			notifyHandlerOnUpdate();
		}
	}

	/**
	 * Adds a state to the state machine.
	 * @param state The state to add. Must not be null.
	 */
	protected void addState(final InputState<E> state) {
		if(state != null) {
			states.add(state);
		}
	}

	/**
	 * Logs (or not) information about the execution of the FSM.
	 * @param log True: logging activated.
	 */
	public void log(final boolean log) {
		if(log) {
			if(logger == null) {
				logger = Logger.getLogger(getClass().getName());
			}
		}else {
			logger = null;
		}
	}

	/**
	 * Reinitialises the FSM.
	 * Remaining events to process are however not clear.
	 * See {@link FSM#fullReinit} for that.
	 */
	public void reinit() {
		if(logger != null) {
			logger.log(Level.INFO, "FSM reinitialised");
		}

		if(currentTimeout != null) {
			currentTimeout.stopTimeout();
		}

		started = false;
		setCurrentState(initState);
		currentTimeout = null;

		if(currentSubFSM != null) {
			currentSubFSM.reinit();
		}
	}

	/**
	 * Reinitialises the FSM.
	 * Compared to {@link FSM#reinit} this method
	 * flushes the remaining events to process.
	 */
	public void fullReinit() {
		synchronized(eventsToProcess) {
			eventsToProcess.clear();
		}
		reinit();

		if(currentSubFSM != null) {
			currentSubFSM.fullReinit();
		}
	}

	/**
	 * Jobs to do when a timeout transition is executed.
	 * Because the timeout transition is based on a separated thread, the job
	 * done by this method must be executed in the UI thread.
	 * UI Platforms must override this method to do that.
	 */
	protected void onTimeout() {
		if(currentTimeout != null) {
			if(logger != null) {
				logger.log(Level.INFO, "Timeout");
			}

			try {
				currentTimeout
					.execute(null)
					.filter(state -> state instanceof OutputState<?>)
					.ifPresent(nextState -> {
						setCurrentState((OutputState<E>) nextState);
						checkTimeoutTransition();
					});
			}catch(final CancelFSMException ignored) {
				// Already processed
			}
		}
	}

	/**
	 * Stops the current timeout transition.
	 */
	protected void stopCurrentTimeout() {
		if(currentTimeout != null) {
			if(logger != null) {
				logger.log(Level.INFO, "Timeout stopped");
			}

			currentTimeout.stopTimeout();
			currentTimeout = null;
		}
	}

	/**
	 * Checks whether the current state has a timeout transition.
	 * If it is the case, the timeout transition is launched.
	 */
	protected void checkTimeoutTransition() {
		currentState
			.getTransitions()
			.stream()
			.filter(tr -> tr instanceof TimeoutTransition)
			.findFirst()
			.map(tr -> (TimeoutTransition<E>) tr)
			.ifPresent(tr -> {
				if(logger != null) {
					logger.log(Level.INFO, "Timeout starting");
				}
				currentTimeout = tr;
				currentTimeout.startTimeout();
			});
	}

	/**
	 * Adds an FSM handler.
	 * @param handler The handler to add.
	 */
	public void addHandler(final FSMHandler handler) {
		if(handler != null) {
			handlers.add(handler);
		}
	}

	/**
	 * Removes the given FSM handler from this FSM.
	 * @param handler The handler to remove.
	 */
	public void removeHandler(final FSMHandler handler) {
		if(handler != null) {
			handlers.remove(handler);
		}
	}

	/**
	 * Notifies handler that the interaction starts.
	 * @throws CancelFSMException If the interaction is cancelled by a handler during the starting step.
	 */
	protected void notifyHandlerOnStart() throws CancelFSMException {
		try {
			for(final FSMHandler handler : handlers) {
				handler.fsmStarts();
			}
		}catch(final CancelFSMException ex) {
			onCancelling();
			throw ex;
		}
	}

	/**
	 * Notifies handler that the interaction updates.
	 * @throws CancelFSMException If the interaction is cancelled by a handler during the updating step.
	 */
	protected void notifyHandlerOnUpdate() throws CancelFSMException {
		try {
			for(final FSMHandler handler : handlers) {
				handler.fsmUpdates();
			}
		}catch(final CancelFSMException ex) {
			onCancelling();
			throw ex;
		}
	}

	/**
	 * Notifies handler that the interaction stops.
	 * @throws CancelFSMException If the interaction is cancelled by a handler during the stopping step.
	 */
	protected void notifyHandlerOnStop() throws CancelFSMException {
		try {
			for(final FSMHandler handler : new ArrayList<>(handlers)) {
				handler.fsmStops();
			}
		}catch(final CancelFSMException ex) {
			onCancelling();
			throw ex;
		}
	}

	/**
	 * Notifies handler that the interaction is cancelled.
	 */
	protected void notifyHandlerOnCancel() {
		new ArrayList<>(handlers).forEach(handler -> handler.fsmCancels());
	}

	/**
	 * @return The set of the states that compose the FSM. This returns a unmodifiable set.
	 */
	public Set<State<E>> getStates() {
		return Collections.unmodifiableSet(states);
	}

	/**
	 * Uninstall the FSM.
	 * Useful for flushing memory.
	 * The FSM must not be used after that.
	 */
	public void uninstall() {
		fullReinit();
		logger = null;
		currentStatePublisher.onComplete();
		startingState = null;
		currentSubFSM = null;
		states.forEach(state -> state.uninstall());
		states.clear();
	}
}
